### DISCUSSION TITLE
Should we maintain consistency between unit__from_node, unit__to_node, and unit__node__node on the data side?

### DISCUSSION THREAD

**Discussion by manuelma** [spine-tools/SpineOpt.jl#465] [Ideas]
I just realized that `add_constraint_unit_pw_heat_rate` doesn't check that `unit__from_node` and `unit__to_node` exists when creating the constraint. This means if one of the flows don't exist, the corresponding term just doesn't appear in the constraint and we end up with unrealistic constraints.

This means unit__from_node, unit__to_node, and unit__node__node must be consistent on the data side. I find this too much work for the modeller who is manipulating data with alternatives. It would be easier if we could rely more on unit__from_node and unit__to_node without needing to make it consistent with unit__node__node.

That means we'd need to write constraints more safely (as opposed to enter data more safely) - but writing constraints is a one-time operation whereas entering data needs to be done multiple times.

**Comment by mihlema** [comment]
I partly agree that the current way can be cumbersome or error-prone. However, I'm not sure I see a way around it if we have multiple inflows/multiple outflows and different relationships. 
For the example above, I think it would definitely be helpful to have a warning in the case that only e.g. `unit__from_node` (or none of the relationships) exist.

**Comment by clizbe** [comment]
I think this is related to our talk in Leuven about data checks and useful warning/error messages. Maybe we should consolidate these ideas somewhere.
